docs/topics/equality.md

```diff
@@ -1,50 +1,123 @@
+Value classes override the `equals()` function to implement structural equality by default.
+
+Similarly, data classes override the `equals()` function to implement structural equality by default. However, if a data class 
+is marked as `final` in the parent class, its `equals()` behavior remains unchanged.
+
+Distinctly, non-data classes (those not declared with the `data` modifier) do not override the 
+`equals()` function by default. Instead, non-data classes implement referential equality behavior inherited from the `Any` class.
+To implement structural equality, non-data classes require a custom equality logic to override the `equals()` function.
+
 To provide a custom equals check implementation, override the
-[`equals(other: Any?): Boolean`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/equals.html) function.
-Functions with the same name and other signatures, like `equals(other: Foo)`, don't affect equality checks with
+[`equals(other: Any?): Boolean`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/equals.html) function:
+
+```kotlin
+class Point(val x: Int, val y: Int) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Point) return false
+
+        // Compares properties for structural equality
+        return this.x == other.x && this.y == other.y
+    }
+}
+```
+
+Functions with the same name and other signatures (like `equals(other: Foo)`) don't affect equality checks with
 the operators `==` and `!=`.
 
 Structural equality has nothing to do with comparison defined by the `Comparable<...>` interface, so only a custom 
 `equals(Any?)` implementation may affect the behavior of the operator. 
 
 ## Referential equality
 
+Referential equality verifies the memory addresses of two objects to determine if they are the same instance.
+
 Referential equality is checked by the `===` operation and its negated counterpart `!==`. `a === b` evaluates to
-true if and only if `a` and `b` point to the same object. For values represented by primitive types at runtime
+true if and only if `a` and `b` point to the same object: 
+
+```kotlin
+fun main() {
+    val a = "Hello"
+    val b = a
+    val c = "world"
+    val d = "world"
+
+    println(a === b)
+    // true
+    println(a === c)
+    // false
+    println(c === d)
+    // true
+
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
+
+For values represented by primitive types at runtime
 (for example, `Int`), the `===` equality check is equivalent to the `==` check.
 
+> For more information about equality in Kotlin/JS, see the [Kotlin/JS](js-interop.md) documentation.
+>
+{type="tip"}
+
 ## Floating-point numbers equality
 
-When an equality check operands are statically known to be `Float` or `Double` (nullable or not), the check follows the 
+Equality checks for `Float` or `Double` operands (nullable or not) follow the 
```

Comment:
It was an important clarification about "statically known". Otherwise it can make an impression that the types of operands are checked to be floating point in runtime.